supervisor/src/bin/supervisor-encore.rs

```diff
@@ -0,0 +1,278 @@
+// It contains the list of available binaries and which services and gateways they implement
+#[derive(serde::Serialize, serde::Deserialize)]
+struct SupervisorConfig {
+    procs: Vec<Proc>,
+}
+
+#[derive(serde::Serialize, serde::Deserialize)]
+struct Proc {
+    id: String,
+    command: Vec<String>,
+    env: Vec<String>,
+    services: Vec<String>,
+    gateways: Vec<String>,
+}
+
+// Process config is the config for a given process
+// It overrides the RuntimeConfig for local service discovery, port allocation
+// and which services and gateways are hosted on this process
+#[derive(serde::Serialize, serde::Deserialize)]
+struct ProcessConfig {
+    local_service_ports: HashMap<String, u16>,
+    hosted_gateways: Vec<String>,
+    hosted_services: Vec<String>,
+}
+
+// RuntimeConfig is a partial version of the config used for GO apps
+// Only hosted services and gateways are parsed as those are the only ones
+// we need to produce ProcessConfigs
+#[derive(serde::Serialize, serde::Deserialize)]
+struct RuntimeConfig {
+    #[serde(default)]
+    hosted_services: Vec<String>,
+    #[serde(default)]
+    gateways: Vec<GatewayConfig>,
+}
+
+#[derive(serde::Serialize, serde::Deserialize)]
+struct GatewayConfig {
+    name: String,
+}
+
+// Create a process config for a given set of services and gateways
+fn create_process_config(
+    services: Vec<String>,
+    gateways: Vec<String>,
+    port: u16,
+    service_ports: &HashMap<String, u16>,
+    cfg: &SupervisorConfig,
+) -> Result<Process, String> {
+    // Append all supervisor environment variables
+    let mut env = std::env::vars().collect::<HashMap<String, String>>();
+
+    // Find a process config that contains all the services and gateways
+    let binary_config = cfg
+        .procs
+        .iter()
+        .find(|p| {
+            (services.iter().all(|s| p.services.contains(s)))
+                && (gateways.iter().all(|g| p.gateways.contains(g)))
+        })
+        .ok_or_else(|| {
+            format!(
+                "No matching proc found for services {:?} gateways {:?}",
+                services, gateways
+            )
+        })?;
+
+    // Add proc-specific environment variables
+    env.extend(binary_config.env.iter().map(|e| {
+        let parts: Vec<&str> = e.splitn(2, '=').collect();
+        (
+            parts[0].to_string(),
+            parts.get(1).unwrap_or(&"").to_string(),
+        )
+    }));
+
+    // Add the port and process config to the environment
+    env.extend(vec![
+        ("PORT".to_string(), port.to_string()),
+        (
+            "ENCORE_PROCESS_CONFIG".to_string(),
+            base64::engine::general_purpose::STANDARD.encode(
+                serde_json::to_string(&ProcessConfig {
+                    hosted_gateways: gateways,
+                    hosted_services: services,
+                    local_service_ports: service_ports.clone(),
+                })
+                .map_err("Failed to serialize ProcessConfig"),
```

Comment:
```suggestion
                .map_err(|e| format!("Failed to serialize ProcessConfig: {}", e))?,
```
Should be more correct, map_err is like map, but you map the error of a result, so the parameter is a closure with the error as its paramter (and then add the `?` at the end to return if there was an error)